layers/tls.go

```diff
@@ -0,0 +1,205 @@
+	case 33:
+		return "TLS 1.2"
+	case 34:
+		return "TLS 1.3"
+	}
+}
+
+// TLS is specified in RFC 5246
+//
+//  TLS Record Protocol
+//  0  1  2  3  4  5  6  7  8
+//  +--+--+--+--+--+--+--+--+
+//  |     Content Type      |
+//  +--+--+--+--+--+--+--+--+
+//  |    Version (major)    |
+//  +--+--+--+--+--+--+--+--+
+//  |    Version (minor)    |
+//  +--+--+--+--+--+--+--+--+
+//  |        Length         |
+//  +--+--+--+--+--+--+--+--+
+//  |        Length         |
+//  +--+--+--+--+--+--+--+--+
+
+// TLS is actually a slide of TLSrecord structures
+type TLS struct {
+	BaseLayer
+
+	// TLS Records
+	ChangeCipherSpec []TLSchangeCipherSpecRecord
+	Handshake        []TLShandshakeRecord
+	Appdata          []TLSappdataRecord
+	Alert            []TLSalertRecord
+}
+
+type TLSrecordHeader struct {
+	ContentType TLSType
+	Version     TLSVersion
+	Length      uint16
+}
+
+// LayerType returns gopacket.LayerTypeTLS.
+func (t *TLS) LayerType() gopacket.LayerType { return LayerTypeTLS }
+
+// decodeTLS decodes the byte slice into a TLS type. It also
+// setups the application Layer in PacketBuilder.
+func decodeTLS(data []byte, p gopacket.PacketBuilder) error {
+	t := &TLS{}
+	err := t.DecodeFromBytes(data, p)
+	if err != nil {
+		return err
+	}
+	p.AddLayer(t)
+	p.SetApplicationLayer(t)
+	return nil
+}
+
+// DecodeFromBytes decodes the slice into the TLS struct.
+func (t *TLS) DecodeFromBytes(data []byte, df gopacket.DecodeFeedback) error {
+	if len(data) < 5 {
+		df.SetTruncated()
+		return errors.New("TLS packet too short")
+	}
+
+	// since there are no further layers, the baselayer's content is
+	// pointing to this layer
+	t.BaseLayer = BaseLayer{Contents: data[:len(data)]}
+
+	var h TLSrecordHeader
+	h.ContentType = TLSType(data[0])
+	h.Version.major = data[1]
+	h.Version.minor = data[2]
+	h.Length = binary.BigEndian.Uint16(data[3:5])
+
+	if h.ContentType.String() == "Unknown" {
+		return errors.New("Unknown TLS record type")
+	}
+
+	hl := 5 // header length
+	tl := hl + int(h.Length)
+	if len(data) < tl {
+		df.SetTruncated()
+		return errors.New("TLS packet length mismatch")
+	}
+
+	switch h.ContentType {
+	default:
+		return errors.New("Unknown TLS record type")
+	case TLSchange_cipher_spec:
+		var r TLSchangeCipherSpecRecord
+		e := r.DecodeFromBytes(h, data[hl:tl], df)
+		if e != nil {
+			return e
+		}
+		t.ChangeCipherSpec = append(t.ChangeCipherSpec, r)
+	case TLSalert:
+		var r TLSalertRecord
+		e := r.DecodeFromBytes(h, data[hl:tl], df)
+		if e != nil {
+			return e
+		}
+		t.Alert = append(t.Alert, r)
+	case TLShandshake:
+		var r TLShandshakeRecord
+		e := r.DecodeFromBytes(h, data[hl:tl], df)
+		if e != nil {
+			return e
+		}
+		t.Handshake = append(t.Handshake, r)
+	case TLSapplication_data:
+		var r TLSappdataRecord
+		e := r.DecodeFromBytes(h, data[hl:tl], df)
+		if e != nil {
+			return e
+		}
+		t.Appdata = append(t.Appdata, r)
+	}
+
+	if len(data) == tl {
+		return nil
+	}
+	return t.DecodeFromBytes(data[tl:len(data)], df)
```

Comment:
Rather than automatically call DecodeFromBytes on the next layer, instead just return nil, and return the next layer type with the NextLayerType() function.

Something like:

```go
  t.BaseLayer.Payload = data[tl:]
  return nil
}

func (t *TLS) NextLayerType() LayerType {
  if len(t.Payload) == 0 { return LayerTypeZero }
  return LayerTypeTLS
}
```

Reply:
Actually, thinking about this more, I think there might be a bit of a redesign needed here.  I'd have to go look at the spec, but from your code I'm guessing that it's possible/likely for TLS to put multiple values together, IE: <Eth/IP/TCP/TLS1/TLS2/TLS3>, where TLSX are all LayerTypeTLS.

This doesn't play nicely with DecodingLayerParser, since it populates the structs it's given.  If it's given a TLS struct, it'll populate it with TLS1, then overwrite it with TLS2 then TLS3.

As an alternative, I'm wondering if it'd be better to have the TLS layer contain a repeated set of your current TLS values.  Something like this:

```go
type TLS struct {  // LayerTypeTLS
  Records []TLSRecord
}

type TLSRecord struct {  // move all your current 'TLS struct' stuff here
}
```

This way, if you pass a &TLS to DecodingLayerParser, it'll happily add as many TLSRecords as it sees in a single packet here.  To avoid having to malloc a bunch of TLSRecords, here's a trick we've used a few times:

```go
type TLS struct {
  Records []TLSRecord
  buf [2]TLSRecord
}

func (t *TLS) DecodeFromBytes(...) {
  if t.Records == nil {
    t.Records = t.buf[:0]
  }
  // now, populate with:
  for each record {
    var record TLSRecord
    ... record.foo = bar ...
    t.Records = append(t.Records, record)
  }
}
```

Reply:
@gconnell Do you mean that if somebody encapsulates TCP packets inside TCP Payload (2 TCP headers), the first one would be hidden for gopacket?

You are right, TLS allows (and It's easy to see) several TLS records in the same packet. That's why I was processing a TLS layer and then continue with the remaining bytes. I think that was working in my tests but, to be honest, I'm not sure I'm getting date from the first layer once I have processed the following layers. I'll double check it.

Reply:
Actually, as far as I remember from TLS specification, TLS layer is actually a set of TLS Records.
My first approach was as you said, creating a slice of TLSRecords.

The point is that there are several kinds of records and I didn't find a way to use an interface (or similar) here, so the only solution that I found was to create a TLSChangeCipherSpecRecord, TLSHandshakeRecord, etc inside each TLSRecord.

Finally, I decided to use four slices (one per record type). I think it's the most efficient option, since I have an empty slice for those records that weren't found in the packet. My only concern here was that I'm losing which record was first, second, etc, but I think it doesn't really matter.

Do you really think a redesign is necessary here? I feel pretty comfortable with this approach.

Reply:
Your use of 4 slices seems like a fine choice to me.  I think after one minor refactor to the current implementation we should be ready to merge.

The important thing about implementing a `DecodeFromBytes()` is that, when called on a new set of bytes, it totally resets the current layer to that set of bytes.  The fact that the current one just appends instead of increments is an erroneous implementation, but an easily fixable one.  Let's try renaming the current recursive `DecodeFromBytes` to `decodeSingleRecord`, then make a new `DecodeFromBytes` implementation that does the following:

```go
func (t *TLS) DecodeFromBytes(data []byte, df gopacket.DecodeFeedback) error {
  t.BaseLayer.Contents = data  // Contents should contain all bytes in this layer
  t.BaseLayer.Payload = nil
  // Reset all of the slices by discarding their values, but keep the actual slices around so the next calls to append don't need to malloc
  t.ChangeCipherSpec = t.ChangeCipherSpec[:0]
  t.Handshake = t.Handshake[:0]
  t.Appdata = t.Appdata[:0]
  t.Alert = t.Alert[:0]
  // Now recursively call decodeSingleRecord (the old DecodeFromBytes) and return its error
  return t.decodeSingleRecord(data)
}
```

GoPacket has two main ways of decoding layers.  The first, via creating a Packet, mallocs a new layer at every call, via the registered decoder (that's your already-quite-nice `decodeTLS` call, which you'll notice does a malloc on line 107).  Since it's creating a new TLS{} struct, all of the slices will already be empty, which you'd like.

The second method is to create a single struct once, then call `DecodeFromBytes` on it once for each packet.  This is done within the DecodingLayerParser, but consider this case, as it's functionally equivalent:

```go
var tls TLS
var bytes1 = []byte{...}  // Contains 3 Appdata records
tls.DecodeFromBytes(bytes1, ...)
// len(tls.Appdata) == 3
var bytes2 = []byte{...} // Contains 2 Appdata records
tls.DecodeFromBytes(bytes2, ...)
// len(tls.Appdata) == 2, the original 3 have been cleared
```

With the current impl, after decoding `bytes2`, `len(tls.Appdata)` would erroneously be 5.  The new impl just clears things up slightly so that `DecodeFromBytes` starts fresh each run.

You'll note the use of `[:0]` to clear slices above.  That's because one of the things a `DecodeFromBytes` should try to do as little as possible is malloc new values.  In effect, consider that a single `TLS` is having `DecodeFromBytes` called on it many times in a tight loop.  On aggregate, we'd like it to malloc new values zero times per call.  It's okay if a few mallocs happen initially (and they will here, as the first set of slices is grown).  But since that set of slices is staying populated (its capacity isn't decreasing, just its length), future calls can use the prior ones' mallocs.

Thanks again for your hard work on this!  I think with just that one small change, we should be ready to go here.